### ISSUE TITLE
Ignore --no-dev option for dependency resolution & lock file generation

### ISSUE THREAD

**Issue by naderman** [composer/composer#5016]
As per https://github.com/composer/composer/issues/5013#issuecomment-193034939 I think the `--no-dev` option should only impact the installation part of the `update` command. The resolution and lock file generation should always include dev packages, to ensure consistent generation of the lock file, regardless of weather `--no-dev` is passed or not. The option should remain, to choose whether dev packages should really be downloaded and installed.

As a side effect this should simplify some of the code in `Composer\Installer`.

**Comment by Seldaek** [comment]
See https://github.com/composer/composer/pull/5224 - it wasn't quite the Easy Pick after all ;)

**Comment by naderman** [comment]
Wow indeed, quite a lot of code!

**Comment by Seldaek** [comment]
Yeah it forced quite a bit of restructuring to avoid code duplication. It doesn't really hurt though to have smaller methods.

**Comment by jpuck** [comment]
Hi @naderman could you please elaborate on what you mean by:

> The resolution and lock file generation should always include dev packages, to ensure consistent generation of the lock file, regardless of weather --no-dev is passed or not.

I'm having trouble wrapping my head around the issue here. After #5224 a lot of [people are having trouble](https://github.com/composer/composer/issues/5594#issuecomment-239737485) with this I think mostly because folks develop in more advanced environments than those deployed.

Thanks!

**Comment by naderman** [comment]
@jpuck That error can only happen if you run install without a lock file, which you should never do in a non-dev environment. Feel free to suggest an update to the docs to make this clearer in a separate issue.
